pkg/kubernetes/deployment/appsv1/v1alpha1/deployment.go

```diff
@@ -0,0 +1,257 @@
+	},
+	// TerminationInProgressPK refer rolloutStatus for predicate IsTerminationInProgress.
+	TerminationInProgressPK: func(d *deploy) string {
+		return fmt.Sprintf("Waiting for deployment rollout to finish: %d old replicas are pending termination",
+			d.object.Status.Replicas-d.object.Status.UpdatedReplicas)
+	},
+	// UpdationInProgressPK refer to rolloutStatus for predicate IsUpdationInProgress.
+	UpdationInProgressPK: func(d *deploy) string {
+		return fmt.Sprintf("Waiting for deployment rollout to finish: %d of %d updated replicas are available",
+			d.object.Status.AvailableReplicas, d.object.Status.UpdatedReplicas)
+	},
+	// NotSyncSpecPK refer to status rolloutStatus for predicate IsNotSyncSpec.
+	NotSyncSpecPK: func(d *deploy) string {
+		return "Waiting for deployment spec update to be observed"
+	},
+}
+
+// rolloutChecks contains a group of predicate it useses predicateKey as key.
+var rolloutChecks = map[predicateKey]predicate{
+	// ProgressDeadlineExceededPK refer to predicate IsProgressDeadlineExceeded.
+	ProgressDeadlineExceededPK: IsProgressDeadlineExceeded(),
+	// OlderReplicaActivePK refer to predicate IsOlderReplicaActive.
+	OlderReplicaActivePK: IsOlderReplicaActive(),
+	// TerminationInProgressPK refer to predicate IsTerminationInProgress.
+	TerminationInProgressPK: IsTerminationInProgress(),
+	// UpdationInProgressPK refer to predicate IsUpdationInProgress.
+	UpdationInProgressPK: IsUpdationInProgress(),
+	// NotSyncSpecPK refer to predicate IsSyncSpec.
+	NotSyncSpecPK: IsNotSyncSpec(),
+}
+
+const (
+	// ProgressDeadlineExceededPK refer to predicate IsProgressDeadlineExceeded.
+	ProgressDeadlineExceededPK predicateKey = "ProgressDeadlineExceeded"
+	// NotSyncSpecPK refer to predicate IsSyncSpec.
+	NotSyncSpecPK predicateKey = "NotSyncSpec"
+	// OlderReplicaActivePK refer to predicate IsOlderReplicaActive.
+	OlderReplicaActivePK predicateKey = "OlderReplicaActive"
+	// TerminationInProgressPK refer to predicate IsTerminationInProgress.
+	TerminationInProgressPK predicateKey = "TerminationInProgress"
+	// UpdationInProgressPK refer to predicate IsUpdationInProgress.
+	UpdationInProgressPK predicateKey = "UpdationInProgress"
+)
+
+// New returns a new instance of deploy meant for deployment
+func New(opts ...deployBuildOption) *deploy {
+	k := &deploy{}
+	for _, o := range opts {
+		o(k)
+	}
+	return k
+}
+
+// WithAPIObject is a deployBuildOption caller can pass deployment schema
+// with this function to create deploy object
+func WithAPIObject(deployment *api_apps_v1.Deployment) deployBuildOption {
+	return func(d *deploy) {
+		d.object = deployment
+	}
+}
+
+// isRollout range over rolloutChecks map and check status of each predicate
+// also it generates status message from rolloutStatuses using predicate key
+func (d *deploy) isRollout() (msg string, ok bool) {
+	for pk, p := range rolloutChecks {
+		if ok = p(d); ok {
+			msg = rolloutStatuses[pk](d)
+			return msg, !ok
+		}
+	}
+	return "", !ok
```

Comment:
function signature level naming or return arguments can create confusion sometimes.
IMO it will be best to avoid naming the return arguments in this particular case.

Reply:
done please check.